src/TLuaInterpreter.cpp

```diff
@@ -11098,58 +11103,78 @@ int TLuaInterpreter::addSupportedTelnetOption( lua_State *L )
-        s++;
-    }
-    if( n > 1 )
-    {
-        if( ! lua_isstring( L, s ) )
-        {
-            lua_pushstring( L, "Echo: wrong argument type" );
-            lua_error( L );
-            return 1;
-        }
-        else
-        {
-            a2 = lua_tostring( L, s );
-        }
-    }
     Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
-    QString txt;
-    QString name;
-    if( n == 1 )
-    {
-        txt = a1.c_str();
-        pHost->mpConsole->buffer.mEchoText = true;
-        pHost->mpConsole->echo( txt );
-        pHost->mpConsole->buffer.mEchoText = false;
+    if (! pHost) {
+        lua_pushstring(L, tr("echo: NULL Host pointer - something is wrong!")
+                       .toUtf8().constData());
+        return lua_error(L);
     }
 
+    QString consoleName;
+    QString displayText;
+    int n = lua_gettop(L);
 
+    if (n > 1) {
+        if (! lua_isstring(L, 1)) {
+            lua_pushstring(L, tr("echo: bad argument #1 type (console name as string, is optional, got %1!)")
+                           .arg(luaL_typename(L, 1))
+                           .toUtf8().constData());
+            return lua_error(L);
+        } else {
+            consoleName = QString::fromUtf8(lua_tostring(L, 1));
+            if (!consoleName.isEmpty()) {
+                if (!consoleName.compare(tr("main","This text is used programmatically, ensure all translations are the same."),
+                                         Qt::CaseSensitive)) {
+
+                    // QString::compare is zero for a match on the "default"
+                    // case so clear the variable - to flag this as the main
+                    // window case - as is the case for an empty string
+                    consoleName.clear();
+                }
+            }
+        }
+    } else if (!n) {
+        // Handle case with NO arguments
+        lua_pushstring(L, tr("echo: bad argument #1 type (text to display as string expected, got nil!)")
+                       .toUtf8().constData());
+        return lua_error(L);
+    }
 
-    else
-    {
-        txt = a2.c_str();
-        name = a1.c_str();
-        mudlet::self()->echoWindow( pHost, name, txt );
+    if (! lua_isstring(L, n)) {
+        lua_pushstring(L, tr("echo: bad argument #%1 type (text to display as string expected, got %2!)")
+                       .arg(n)
+                       .arg(luaL_typename(L, n))
+                       .toUtf8().constData()); // Cannot combine the 2 args, is ambiguous
+        return lua_error(L);
+    } else {
+        displayText = QString::fromUtf8(lua_tostring(L, n));
     }
 
-    return 0;
+    if (consoleName.isEmpty()) {
+        pHost->mpConsole->buffer.mEchoText = true;
+        pHost->mpConsole->echo(displayText);
+        pHost->mpConsole->buffer.mEchoText = false;
+        // Writing to the main window must always succeed, but for consistent
+        // results, we now return a true for that
+        lua_pushboolean(L, true);
```

Comment:
I don't think we should return true here because it implies there is a case it'll fail - implies people have to add error checking. @keneanung ?